ddtrace/internal/_encoding.pyx

```diff
@@ -408,3 +578,86 @@ cdef class MsgpackEncoder(MsgpackEncoderBase):
                 if ret != 0: return ret
 
         return ret
+
+
+cdef class MsgpackEncoderV05(MsgpackEncoderBase):
+    cdef object _st
+
+    def __cinit__(self, max_size, max_item_size):
+        self._st = MsgpackStringTable(max_size)
+
+    cpdef flush(self):
+        try:
+            self._st.append_raw(PyLong_FromLong(<long> self.get_buffer()), <Py_ssize_t> super(MsgpackEncoderV05, self).size)
+            return self._st.flush()
+        finally:
+            # Reset the buffer.
+            self.pk.length = 5
+            self._count = 0
+
+    @property
+    def size(self):
+        """Return the size in bytes of the encoder buffer."""
+        return self._st.size + super(MsgpackEncoderV05, self).size
+
+    cpdef put(self, list trace):
+        try:
+            self._st.savepoint()
+            super(MsgpackEncoderV05, self).put(trace)
+        except Exception:
+            self._st.rollback()
+            raise
+
+    cdef inline int _pack_string(self, object string):
+        return msgpack_pack_uint32(&self.pk, self._st.index(string))
+
+    cpdef pack_span(self, object span, object dd_origin):
```

Comment:
I wonder if we should pre-insert dd_origin into the string table once before `pack_span`, then we can just pass an index id around to insert.

Reply:
👍  good idea!